### PR TITLE
Add #381 [WIP]

### PR DESCRIPTION
**Pull request by turtles2**
This will add #381 . Its current state its a proof of concept it needs the following things which I will work on.  

- [ ] Extend to all Models 
- [ ] Add Tests for added functions 
- [ ] Modify Tests to Work with new URL structure
- [ ] Allow both old and new URLs to work
- [ ] Add HashIDs to URLs that don't use binding

### CODE REVIEW

File: app/Helpers/ID_hasher.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace App\Helpers;
+
+use Vinkla\Hashids\Facades\Hashids;
+
+class ID_hasher
```
Comment (asbiin) [inline app/Helpers/ID_hasher.php]: MINOR Rename class "ID_hasher" to match the regular expression ^[A-Z][a-zA-Z0-9]*$. [rule](https://sonarcloud.io/coding_rules#rule_key=php%3AS101)

File: app/Providers/RouteServiceProvider.php [hunk 2]
```diff
@@ -39,8 +40,11 @@ public function boot()
     {
         parent::boot();
 
-        Route::bind('contact', function ($value) {
+        $ID_hasher = new ID_hasher();
+
+        Route::bind('contact', function ($value) use ($ID_hasher){
```
Comment (asbiin) [inline app/Providers/RouteServiceProvider.php]: MINOR Define a constant instead of duplicating this literal "contact" 11 times. [rule](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)

**Comment by asbiin** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by djaiss** [comment]
@turtles2 instead of writing `Add #381` in the title, could you write a description of what the PR does? It'll be easier. Thanks! And great for the feature you are trying to implement!

File: app/Providers/RouteServiceProvider.php [hunk 3]
```diff
@@ -56,13 +60,15 @@ public function boot()
                 ->firstOrFail();
         });
 
-        Route::bind('activity', function ($value, $route) {
+        Route::bind('activity', function ($value, $route) use ($ID_hasher) {
```
Comment (asbiin) [inline app/Providers/RouteServiceProvider.php]: MAJOR Remove the unused function parameter "$route". [rule](https://sonarcloud.io/coding_rules#rule_key=php%3AS1172)

**Comment by asbiin** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by turtles2** [comment]
@djaiss I am unsure on the best way to handle the existing tests. Many of them right now (18) use the model ID in there URL structure that will continue to work. But do we want to change them to use the hashid instead? At first I was thinking that we do but now I think it would be best to leave them as is and only test hashid for the sake of testing hashids.

File: app/Providers/RouteServiceProvider.php [hunk 4]
```diff
@@ -39,8 +40,12 @@ public function boot()
     {
         parent::boot();
 
-        Route::bind('contact', function ($value) {
+        $ID_hasher = new ID_hasher();
+
+        Route::bind('contact', function ($value) use ($ID_hasher) {
```
Comment (asbiin) [inline app/Providers/RouteServiceProvider.php]: MINOR Define a constant instead of duplicating this literal "contact" 11 times. [rule](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)

Comment (asbiin) [inline app/Providers/RouteServiceProvider.php]: MINOR Define a constant instead of duplicating this literal "contact" 11 times. [rule](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)

**Comment by djaiss** [comment]
@turtles2 ping me when you think you're done :-D

**Comment by turtles2** [comment]
@djaiss I believe it's what ready to pull in expect for the following which I asked previously and am awaiting a response. I also need to add a change log entry but that's trivial. 

> I am unsure on the best way to handle the existing tests. Many of them right now (18) use the model ID in there URL structure that will continue to work. But do we want to change them to use the hashid instead? At first I was thinking that we do but now I think it would be best to leave them as is and only test hashid for the sake of testing hashids.

File: .env.example [hunk 5]
```diff
@@ -12,6 +12,9 @@ APP_DEBUG=false
 # Use `php artisan key:generate` to generate a random key.
 APP_KEY=ChangeMeBy32KeyLengthOrGenerated
```
Comment (djaiss) [inline .env.example]: We need to put those a comment above these two new parameters to explain what it does.

I suggest
```
# Prevent information leakage by referring to contacts with hashIds instead of
# the actual IDs used in the database.
```

File: .env.example [hunk 6]
```diff
@@ -12,6 +12,9 @@ APP_DEBUG=false
...
 
+HASH_SALT=SALT
```
Comment (djaiss) [inline .env.example]: Change this by 
```
HASH_SALT=ChangeMeBy20+KeyLength
```

File: app/Helpers/ID_hasher.php (hunk 1 above)
Comment (djaiss) [reply app/Helpers/ID_hasher.php]: The name `ID_hasher` is not [PSR1 compliant](http://www.php-fig.org/psr/psr-1/).

Please change it everywhere with `idHasher`.

File: app/Helpers/ID_hasher.php [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public function encode_id($id)
```
Comment (djaiss) [inline app/Helpers/ID_hasher.php]: `encode_id` → `encodeId` for PSR1 compliance.

File: app/Helpers/ID_hasher.php [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        return 'h'.Hashids::encode($id);
+    }
+
+    public function decode_id($hash)
```
Comment (djaiss) [inline app/Helpers/ID_hasher.php]: `decode_id` → `decodeId`

File: docs/contribute/contribute.md [hunk 9]
```diff
@@ -194,6 +195,9 @@ Reminders are generated and sent using an Artisan command `monica:sendnotificati
 Monica calculates every night (ie once per day) a set of metrics to help you understand how the instance is being used by users. That will also allow to measure growth over time.
 
 Statistics are generated by the Artisan command `monica:calculatestatistics` every night at midnight and this cron is defined in `app/console/Kernel.php`.
+## Routing
```
Comment (djaiss) [inline docs/contribute/contribute.md]: Please add a line above for better readability.

Also, the text is poorly written.

Please change it to

```
When adding new routes that inject Models, you need to either 
* use Laravel's route helper to inject the model (this is the preferred way)
* if for some reason you can't call the route helper instead of putting the ID in the URL, call `hashID()` on the model you are trying to inject this is to ensure that URLs are not leaking information. 

Also, if this is the first time that Model is being used in a route, ensure that it uses the `hasher` trait.
```

**Comment by turtles2** [comment]
@djaiss Thanks for pointing out where it's not being done, the project has a lot of URLs so it's easy for me to miss them. I will work on adding those along with doing the other changes you requested. The semester is just picking up steam so it might take me a little bit. 

I chose to do it on the fly for a few reasons. The first of which is that if we store them in the DB it would require the salt to never change or be stored with it. Right now if the salt changes for whatever reason all the users would need to do is start a new session. If we stored in the DB we would either need to rehash the entire DB or have the overhead of storing the slat. Secondly by not storing in the DB we make it easier to add HashIds to another model. Right now if you want to add it to a model all you need to do is add the trait to said model and add whatever routes you want. If we were to store in the DB we would need a migration to add the column plus a script to generate the IDs for the existing records. Ultimately my goal is for this to be added to the project with minimal impact for existing installs.  Right now the only impact is the installation of dependencies and required PHP extensions.  From my testing it had minimal impact on performance. 

tl;dr It is done on the fly to make implementation for all routes easier and to minimize impact.

File: .env.example (hunk 5 above)
Comment (turtles2) [reply .env.example]: Done

File: .env.example (hunk 6 above)
Comment (turtles2) [reply .env.example]: Done

File: docs/contribute/contribute.md (hunk 9 above)
Comment (turtles2) [reply docs/contribute/contribute.md]: Done

**Comment by turtles2** [comment]
@djaiss I hope to have this done and pulled in this week, Is there anything else that needs to be done besides what has already been mentioned?

**Comment by turtles2** [comment]
@djaiss As far as I can tell I have made all the requested changes. When you get a chance please let me know what steps we need to take next.

**Comment by djaiss** [comment]
Alright. Thanks so much for your work.

There is one more thing 😀

When we update the contact, we get redirected to `people/ID` and not `people/hashID`. This also happens everywhere we edit something (therefore leaving the contact view) and save the edit.

**Comment by turtles2** [comment]
@djaiss I have added them to the edit and create returns as well as to some some other areas that were missing them. 

Please review again when you get a chance.

**Comment by turtles2** [comment]
@djaiss I merged in the changes that were made to the upstream.  I am unable to recompile assets at this time but if this is not merged in by than I will do so this weekend.

**Comment by djaiss** [comment]
This is my priority for this weekend!

**Comment by turtles2** [comment]
@djaiss Sounds good, I plan to login into my server tonight to recompile the assets and push those changes up.

File: public/mix-manifest.json [hunk 10]
```diff
@@ -1,10 +1,6 @@
...
-    "/css/stripe.css": "/css/stripe.css?id=956554a2e96db30fafa3",
-    "/js/app.js.map": "/js/app.js.map?id=e62443022dbabf2c461f",
-    "/css/app.css.map": "/css/app.css.map?id=a1585d15bdf5306642f7",
-    "/css/stripe.css.map": "/css/stripe.css.map?id=80d7af9d49b0922d35e6",
-    "/js/stripe.js": "/js/stripe.js?id=55d823b61fc4f7e8b1ab",
-    "/js/stripe.js.map": "/js/stripe.js.map?id=4f29019f2c2bda216ee1"
+    "/js/app.js": "/js/app.js?id=f6844194a630b44f9e48",
+    "/css/app.css": "/css/app.css?id=2b58773eb3b4f2c35795",
+    "/css/stripe.css": "/css/stripe.css?id=c16ddc73ba6af94a5bfa",
+    "/js/stripe.js": "/js/stripe.js?id=b3b3f79ea16865ee5107"
```
Comment (asbiin) [inline public/mix-manifest.json]: You want to compile assets with `npm run production`, not in `dev` mode.

File: .env.example [hunk 11]
```diff
@@ -12,6 +12,12 @@ APP_DEBUG=false
 # Use `php artisan key:generate` to generate a random key.
 APP_KEY=ChangeMeBy32KeyLengthOrGenerated
 
+# Prevent information leakage by referring to IDs with hashIds instead of
+# the actual IDs used in the database.
+
```
[the remark this answers is not in the document]
Comment (turtles2) [reply .env.example]: Done

File: app/Helpers/idHasher.php [hunk 12]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+namespace App\Helpers;
+
+use Vinkla\Hashids\Facades\Hashids;
+
+class idHasher
```
[the remark this answers is not in the document]
Comment (turtles2) [reply app/Helpers/idHasher.php]: Done

File: composer.json [hunk 13]
```diff
@@ -28,6 +28,11 @@
         "pragmarx/recovery": "^0.1",
         "predis/predis": "^1.1",
         "sabre/vobject": "^4.1",
+        "vinkla/hashids": "^3.3",
+        "paragonie/constant_time_encoding": "^2.2",
```
[the remark this answers is not in the document]
Comment (turtles2) [reply composer.json]: Done?

File: public/mix-manifest.json (hunk 10 above)
Comment (turtles2) [reply public/mix-manifest.json]: Done

**Comment by turtles2** [comment]
@djaiss @asbiin The requested changes have been made

**Comment by asbiin** [comment]
That's good for me.

I have however 2 improvements proposal:
- Use a service provider to create a singleton for IdHasher, instead of creating a new one
[idhasher_service_provider.zip](https://github.com/monicahq/monica/files/1863145/idhasher_service_provider.zip)
(or we could use static fonction for encodeId/decodeId, but with the use of `config('hashids.default_prefix')` it's better that way)
- Use hash for every requests: there is still some `contact->id` in resources/views that could be "hashized"

Tanks !

**Comment by turtles2** [comment]
@asbiin I will honest I lack a solid understanding of service providers (I would dig into them but I am preoccupied with School Ocaml and C at the same time require a lot of thinking) it seems looking at the diff you already have thought of how to implement it. I am not partial to the way I implemented this, I just used a method that made sense at the time.  The one thing the diff missed was the unit tests for the class. 

As for the missing views I am not sure how best to approach to find them all besides brute forcing as there are many instances `contact->id` that do not pertain to views which maks find and replace difficult. With any luck I will have some time to tomorrow and I can rexamine the files and see if I can find them.

**Comment by turtles2** [comment]
@asbiin I would be good with that.

**Comment by asbiin** [comment]
@turtles2 OK I've added the ServiceProvider. One file was indeed missed from my zipped patch !

**Comment by turtles2** [comment]
@djaiss @asbiin I have added the hashes to more views/routes I believe this should get most/all of them. Let me know if we need to make any additional changes.

**Comment by turtles2** [comment]
@djaiss I have added the hashIDs to the new relationships but I am getting an error "global_relationship_form_new_contact is not defined" on the editing page. I not sure why I am getting this. Besides this error it is ready for merging.

**Comment by djaiss** [comment]
@turtles2 yes, this error is due to this variable missing in `app.js`.

To fix it, until a proper fix arrives, you can do this at the bottom of `app.js`:

```javascript
loadLanguageAsync(window.Laravel.locale, true).then((lang) => {

    // the Vue appplication
    me.app = new Vue({
      i18n,
      data: {
        activities_description_show: false,
        reminders_frequency: 'once',
        accept_invite_user: false,
        date_met_the_contact: 'known',
        global_relationship_form_new_contact: true,   ←-----------------HERE
      },
      methods: {
      },
      mounted: function() {

        // required modules
        require('./tags');
        require('./search');
        require('./contacts');

      }
    }).$mount('#app');

    return app;
});
```

and recompile the assets (`npm run prod`) and it should work.

File: resources/views/people/debt/index.blade.php [hunk 14]
```diff
@@ -46,7 +46,7 @@
           @endif
         </div>
         <div class="table-cell list-actions">
-          <a href="{{ route('people.debt.edit', ['people' => $contact->id, 'debtId' => $debt->id]) }}">
+          <a href="{{ route('people.debt.edit', [$contact, $debt]) }}">
```
Comment (asbiin) [inline resources/views/people/debt/index.blade.php]: Why not keep the named attributes here ?

Comment (turtles2) [reply resources/views/people/debt/index.blade.php]: @asbiin I would have to test it to be sure but it's one of two reasons. Either there was some quirk in laravels route binding that required it, or I got carried away with find and replace. Looking at I would  lean towards the first one.

Comment (asbiin) [reply resources/views/people/debt/index.blade.php]: @turtles2 indeed ! The previous version was obviously buggy. Thanks !

**Comment by asbiin** [comment]
We have just switched to Laravel 5.6, so I've updated vinkla/hashids dependency to '^5.0'

File: app/Helpers/IdHasher.php [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public function encodeId($id)
+    {
+        return $this->prefix.Hashids::encode($id);
+    }
+
+    public function decodeId($hash)
+    {
+        if (str_contains($hash, $this->prefix)) {
```
[the remark this answers is not in the document]
Comment (turtles2) [reply app/Helpers/IdHasher.php]: @asbiin Yes it would be. I was not aware of that function.

**Comment by asbiin** [comment]
Contact deletion seems to be broken, can you confirm ?

**Comment by turtles2** [comment]
I will take a look tonight and if that's the case hopefully fix it as well.

**Comment by asbiin** [comment]
@turtles2 It was related to another issue, fixed in #1114
